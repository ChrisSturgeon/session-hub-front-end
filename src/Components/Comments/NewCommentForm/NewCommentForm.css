.new-comment-form {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.new-comment-form button {
  background-color: var(--medium-blue);
  color: white;
  border-radius: 5px;
  border: none;
  padding: 0.25em;
  margin: 0.5em 0;
}

.new-comment-form textarea {
  min-width: 100%;
  max-width: 100%;
  height: 150px;
}

.new-comment-form .character-count {
  text-align: right;
}

@media only screen and (min-width: 768px) {
  .new-comment-form button {
    width: 200px;
    align-self: flex-end;
  }
}
