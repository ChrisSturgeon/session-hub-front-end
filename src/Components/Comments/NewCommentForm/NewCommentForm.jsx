import './NewCommentForm.css';
import { useState, useEffect, useContext } from 'react';
import { UserContext } from '../../../App';
import { useParams, useNavigate } from 'react-router-dom';
import { APIURL } from '../../../api';
import { motion, AnimatePresence } from 'framer-motion';

const expandDown = {
  initial: {
    height: 0,
    opacity: 0,
  },
  animate: {
    height: 'auto',
    opacity: 1,
    transition: {
      height: {
        duration: 0.2,
      },
      opacity: {
        duration: 0.3,
      },
    },
  },
  exit: {
    height: 0,
    opacity: 0,
    transition: {
      height: {
        duration: 0.5,
      },
      opacity: {
        duration: 0.2,
      },
    },
  },
};

export default function NewCommentForm() {
  const navigate = useNavigate();
  const user = useContext(UserContext);
  const { sessionID } = useParams();
  const [show, setShow] = useState(true);
  const [text, setText] = useState('');
  const [characterCount, setCharacterCount] = useState(0);

  function handleChange(event) {
    setText((prev) => event.target.value);
  }

  useEffect(() => {
    setCharacterCount((prev) => text.length);
  }, [text]);

  async function handleSubmit(event) {
    event.preventDefault();
    const sessionURL = `${APIURL}/comments/${sessionID}`;

    try {
      const commentObj = {
        date: new Date(),
        text: text,
      };
      const response = await fetch(sessionURL, {
        method: 'POST',
        mode: 'cors',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `bearer ${window.localStorage.getItem('JWT')}`,
        },
        body: JSON.stringify(commentObj),
      });

      const data = await response.json();

      if (response.status === 201) {
        console.log(data.message);
        setShow(false);
        return;
      }

      if (response.status === 404) {
        console.log(data.message);
        return;
      }
    } catch (err) {
      console.log(err);
    }
  }

  return (
    <>
      <AnimatePresence mode="wait">
        {show && (
          <motion.form
            initial={expandDown.initial}
            animate={expandDown.animate}
            exit={expandDown.exit}
            onSubmit={handleSubmit}
            className="new-comment-form"
          >
            <label>New Comment as {user.username}</label>
            <textarea onChange={handleChange} value={text}></textarea>
            <div className="character-count">{characterCount} / 1500 </div>
            <button>Submit Comment</button>
          </motion.form>
        )}
      </AnimatePresence>

      {!show && <div>Thanks for your comment!</div>}
    </>
  );
}
