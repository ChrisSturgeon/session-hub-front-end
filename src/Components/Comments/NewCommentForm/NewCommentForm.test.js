import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import NewCommentForm from './NewCommentForm';

describe('New Comment Form', () => {
  it('Displays form', () => {
    render(<NewCommentForm />);
    expect(screen.getByRole('form')).toBeInTheDocument();
  });

  it('Matches snapshot', () => {
    const { container } = render(<NewCommentForm />);
    expect(container).toMatchSnapshot();
  });

  it('Displays add button', () => {
    render(<NewCommentForm />);
    expect(
      screen.getByRole('button', { name: /Add Comment/i })
    ).toBeInTheDocument();
  });

  it('Displays textarea', () => {
    render(<NewCommentForm />);
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('Textarea has placeholder', () => {
    render(<NewCommentForm />);
    expect(screen.getByRole('textbox')).toHaveAttribute(
      'placeholder',
      'Leave a comment!'
    );
  });

  it('Renders with zero character count', () => {
    render(<NewCommentForm />);
    expect(screen.getByTestId('character-count')).toHaveTextContent('0 / 1500');
  });

  it('Textarea updates on typing', () => {
    render(<NewCommentForm />);
    const textarea = screen.getByRole('textbox');
    userEvent.type(textarea, 'Session Hub');
    expect(textarea).toHaveTextContent('Session Hub');
  });

  it('Character count updates on typing', () => {
    render(<NewCommentForm />);
    const textarea = screen.getByRole('textbox');
    userEvent.type(textarea, 'Session Hub');
    expect(screen.getByTestId('character-count')).toHaveTextContent(
      '11 / 1500'
    );
  });

  it('Displays minimum character length error', () => {
    render(<NewCommentForm />);
    const textarea = screen.getByRole('textbox');
    userEvent.type(textarea, 'H');
    const submitButton = screen.getByRole('button');
    userEvent.click(submitButton);
    expect(screen.getByTestId('validation-error')).toBeInTheDocument();
  });
});
