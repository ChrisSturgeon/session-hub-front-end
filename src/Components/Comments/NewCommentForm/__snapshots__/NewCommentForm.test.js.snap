// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Comment Form Matches snapshot 1`] = `
<div>
  <form
    class="new-comment-form"
    name="new-comment-form"
    style="height: auto; opacity: 1;"
  >
    <textarea
      placeholder="Leave a comment!"
    />
    <div
      class="character-count"
      data-testid="character-count"
    >
      0
       / 1500
       
    </div>
    <button>
      Add Comment
    </button>
  </form>
</div>
`;
